### PR TITLE
postgresql_owner - Add support at more obj_type

### LINKED ISSUE #482: feature request: support more obj_type in postgresql_owner
Currently, the _postgresql_owner_ module supports altering the owner of these kinds of things:

- `database`
- `function`
- `matview`
- `sequence`
- `schema`
- `table`
- `tablespace`
- `view`

But there are `ALTER $kind $thing OWNER TO $owner` commands (checked for postgres 15) for more unusual kinds of things:

- `aggregate`
- `collation`
- `conversion`
- `domain`
- `event trigger`
- `foreign data wrapper`
- `foreign table`
- `function`
- `language`
- `large object`
- `operator`
- `operator class`
- `operator family`
- `procedure`
- `publication`
- `routine`
- `server`
- `statistics`
- `subscription`
- `tablespace`
- `text search configuration`
- `text search dictionary`
- `type`

It would be nice to have them integrated in the _postgresql_owner_ module (available as `obj_type`).
I'm mostly interested in this subset of the currently-not-supported object kinds:

- `aggregate`
- `collation`
- `conversion`
- `domain`
- `function`
- `operator`
- `operator class`
- `operator family`
- `procedure`
- `routine`
- `type`

### PR DESCRIPTION
**Pull request by RealGreenDragon**
##### SUMMARY

Partially fixes #482

My PR adds in postgresql_owner the support (with all necessary tests) at following object types:

- `procedure`
- `type`
- `aggregate`
- `routine`
- `language`
- `domain`
- `collation`
- `conversion`
- `text search configuration`
- `text search dictionary`
- `foreign data wrapper`
- `server`
- `foreign table`
- `event trigger`
- `large object`
- `publication`
- `statistics`

Following object types cited in issue are already present:

- `function`
- `tablespace`

Following object types cited in issue are not been added:

- `operator` (complex and few time)
- `operator class` (complex and few time)
- `operator family` (complex and few time)
- `subscription` (use `postgresql_subscription` module for it)

Note that (as written in module notes) following object types have a limited implementation (Function Overloading not supported):

- `function`
- `procedure`
- `aggregate`
- `routine`

Other fixes:

- Various tests fixes
- Added server version check

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

postgresql_owner

### CODE REVIEW

**Comment by RealGreenDragon** [comment]
Ready for review

File: plugins/modules/postgresql_owner.py [hunk 1]
```diff
@@ -292,7 +366,11 @@ def __is_owner(self):
                      "WHERE d.datname = %(obj_name)s "
                      "AND r.rolname = %(role)s")
 
-        elif self.obj_type == 'function':
+        elif self.obj_type in ('aggregate', 'function', 'routine', 'procedure'):
+            if self.obj_type == 'routine' and self.pg_version < 110000:
+                raise Error("PostgreSQL version must be >= 11 for obj_type=routine. Exit")
```
Comment (Andersson007) [inline plugins/modules/postgresql_owner.py:371]:
```suggestion
                self.module.fail_json(msg="PostgreSQL version must be >= 11 for obj_type=routine.")
```
this is a standard way of handling such stuff, i.e. all interactions with a user are done through the `module` object . Relevant for all similar cases. I would also suggest removing `Exit` from the messages.

**Review by Andersson007 — CHANGES_REQUESTED** [review]
@RealGreenDragon thanks for the great improvement!

File: plugins/modules/postgresql_owner.py [hunk 2]
```diff
@@ -74,10 +80,20 @@
     type: bool
     default: true
     version_added: '0.2.0'
+
+notes:
+- Function Overloading is not supported so when I(obj_type) is C(aggregate), C(function), C(routine), or C(procedure)
+  I(obj_name) is considered the only object of same type with this name.
+- Despite Function Overloading is not supported, when I(obj_type=aggregate) I(obj_name) must contain also aggregate
```
Comment (Andersson007) [inline plugins/modules/postgresql_owner.py:87]:
```suggestion
- Function Overloading is not supported so when I(obj_type) is C(aggregate), C(function), C(routine), or C(procedure)
  I(obj_name) is considered the only object of same type with this name.
- Despite Function Overloading is not supported, when I(obj_type=aggregate) I(obj_name) must contain also aggregate
```
Good-looking readable notes split into several lines.

Comment (RealGreenDragon) [reply plugins/modules/postgresql_owner.py:87]: ok

File: plugins/modules/postgresql_owner.py (hunk 1 above)
Comment (RealGreenDragon) [reply plugins/modules/postgresql_owner.py]: Good advice, fixed.

File: plugins/modules/postgresql_owner.py [hunk 3]
```diff
@@ -384,14 +659,18 @@ def __role_exists(self, role):
 # Module execution.
 #
 
+VALID_OBJ_TYPES = ('aggregate', 'collation', 'conversion', 'database', 'domain', 'event_trigger', 'foreign_data_wrapper',
+                   'foreign_table', 'function', 'language', 'large_object', 'matview', 'procedure', 'publication',
+                   'routine', 'schema', 'sequence', 'server', 'statistics', 'table', 'tablespace', 'text_search_configuration',
+                   'text_search_dictionary', 'type', 'view')
```
Comment (Andersson007) [inline plugins/modules/postgresql_owner.py:665]: i think we should put this constant right after the packages-import section

Comment (RealGreenDragon) [reply plugins/modules/postgresql_owner.py]: Done

**Review by Andersson007 — APPROVED** [review]

File: plugins/modules/postgresql_owner.py [hunk 4]
```diff
@@ -292,7 +368,11 @@ def __is_owner(self):
                      "WHERE d.datname = %(obj_name)s "
                      "AND r.rolname = %(role)s")
 
-        elif self.obj_type == 'function':
+        elif self.obj_type in ('aggregate', 'function', 'routine', 'procedure'):
+            if self.obj_type == 'routine' and self.pg_version < 110000:
+                self.module.fail_json(msg="PostgreSQL version must be >= 11 for obj_type=routine.")
```
Comment (Andersson007) [inline plugins/modules/postgresql_owner.py:373]:
```suggestion
            if self.obj_type == 'routine' and self.pg_version < 110000:
                self.module.fail_json(msg="PostgreSQL version must be >= 11 for obj_type=routine.")
```
This is just a FYI, no need to do it right now. Let's merge this PR if no objections.
In a future PR we could avoid the code repetitions. The ides is to create a matrix (but it maybe stupid and will not work but theoretically it should):
```
OBJ_TYPES_PG_VER_MAP = {
    'routine': 110000,
    ...
    'typeN': NUM,
}
```
Then in one place (maybe in the class in the initialization step) we could check it by calling a dedicated function containing:
```
if OBJ_TYPES_PG_VER_MAP['self.obj_type'] > self.pg_version:
    self.module.fail_json(msg="PostgreSQL version must be >= %s for obj_type=%s" % (OBJ_TYPES_PG_VER_MAP['self.obj_type'], self.obj_type))
```
For the old parameters present in all the version the collection supports we could just put `0`)

In `choices`, you can extract the keys using `OBJ_TYPES_PG_VER_MAP.keys()` IIRC the syntax

The idea for future PR

**Review by Andersson007 — APPROVED** [review]
@RealGreenDragon LGTM
(i've put an idea for another PR)

File: plugins/modules/postgresql_owner.py [hunk 5]
```diff
@@ -74,10 +80,20 @@
     type: bool
     default: true
     version_added: '0.2.0'
+
+notes:
+- Function Overloading is not supported so when I(obj_type) is C(aggregate), C(function), C(routine), or C(procedure)
+  I(obj_name) is considered the only object of same type with this name.
+- Despite Function Overloading is not supported, when I(obj_type=aggregate) I(obj_name) must contain also aggregate
+  signature because it is required by SQL syntax.
+- I(new_owner) must be asuperuser if I(obj_type) is C(event_type) or C(foreign_data_wrapper).
```
Comment (tcraxs) [inline plugins/modules/postgresql_owner.py:89]:
```suggestion
- I(new_owner) must be a superuser if I(obj_type) is C(event_type) or C(foreign_data_wrapper).
```

**Review by tcraxs — APPROVED** [review]
LGTM just found a typo

**Review by tcraxs — APPROVED** [review]

**Comment by Andersson007** [comment]
@hunleyd ?

**Review by hunleyd — APPROVED** [review]

**Comment by Andersson007** [comment]
@RealGreenDragon thanks for the contribution!
@tcraxs @hunleyd thanks for reviewing and merging!

**Pull request merged**